linux-gpib-user/doc/man/man8/gpib_config.8
'\" t
.\"     Title: gpib_config
.\"    Author: Frank Mori Hess
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 10/04/2025
.\"    Manual: 	Configuration
.\"    Source: linux-gpib 4.3.7
.\"  Language: English
.\"
.TH "GPIB_CONFIG" "8" "10/04/2025" "linux-gpib 4.3.7" "Configuration"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gpib_config \- GPIB administration program
.SH "SYNOPSIS"
.HP \w'\fBgpib_config\fR\ 'u
\fBgpib_config\fR [\-\-minor\ \fInumber\fR]
.HP \w'\fBgpib_config\fR\ 'u
\fBgpib_config\fR [\-\-board\-type\ \fIboard_type\fR] [\-\-device\-file\ \fIfile_path\fR] [\-\-dma\ \fInumber\fR] [\-\-file\ \fIfile_path\fR] [\-\-help] [\-\-iobase\ \fInumber\fR] [\-\-ifc] [\-\-no\-ifc] [\-\-init\-data\ \fIfile_path\fR] [\-\-irq\ \fInumber\fR] [\-\-minor\ \fInumber\fR] [\-\-offline] [\-\-pad\ \fInumber\fR] [\-\-pci\-bus\ \fInumber\fR] [\-\-pci\-slot\ \fInumber\fR] [\-\-sad\ \fInumber\fR] [\-\-serial\-number\ \fIserial_number\fR] [\-\-sre] [\-\-no\-sre] [\-\-sysfs\-device\-path\ \fIsysfs_device_path\fR] [\-\-system\-controller] [\-\-no\-system\-controller] [\-\-version]
.SH "DESCRIPTION"
.PP
gpib_config must be run after the kernel driver module for a GPIB interface board is loaded\&. It performs configuration of driver settings that cannot be performed by libgpib at runtime\&. This includes configuration which requires root privilege (for example, setting the base address or irq of a board), and configuration which should only be performed once and not automatically redone every time a program using libgpib is run (for example, setting the board\*(Aqs GPIB address)\&.
.PP
The board to be configured by gpib_config is selected by the
\fB\-\-minor\fR
option\&. By default, the board settings are read from the
gpib\&.conf
configuration file\&. However, individual settings can be overiden by use of command\-line options (see below)\&.
.SH "OPTIONS"
.PP
\fB\-t, \-\-board\-type \fR\fB\fIboard_type\fR\fR
.PP
Set board type to
\fIboard_type\fR\&.
.PP
\fB\-c, \-\-device\-file \fR\fB\fIfile_path\fR\fR
.PP
Specify character device file path for the board\&. This can be used as an alternative to the \-\-minor option\&.
.PP
\fB\-d, \-\-dma \fR\fB\fInumber\fR\fR
.PP
Specify isa dma channel
\fInumber\fR
for boards without plug\-and\-play cabability\&.
.PP
\fB\-f, \-\-file \fR\fB\fIfile_path\fR\fR
.PP
Specify file path for configuration file\&. The values in the configuration file will be used as defaults for unspecified options for the minor being configured\&. If this option is not specified, then the value of the IB_CONFIG environment variable will be used if it is set\&. Otherwise, the default configuration file is "sysconfdir/gpib\&.conf" where sysconfdir is specified when the library is configured before compilation (it is usually /etc or /usr/local/etc)\&. Note: The configuration file read by the library is not affected by this option\&. It will either use the value of the IB_CONFIG environment variable, if it is set, or the default configuration file\&. The configuration file read by the library is used to resolve the \*(Aqname\*(Aq in ibfind() calls\&.
.PP
\fB\-h, \-\-help\fR
.PP
Print help on options and exit\&.
.PP
\fB\-I, \-\-init\-data \fR\fB\fIfile_path\fR\fR
.PP
Upload binary initialization data (firmware) from
\fIfile_path\fR
to board\&.
.PP
\fB\-\-[no\-]ifc\fR
.PP
Perform (or not) interface clear after bringing board online\&. Default is
\fB\-\-ifc\fR\&.
.PP
\fB\-b, \-\-iobase \fR\fB\fInumber\fR\fR
.PP
Set io base address to
\fInumber\fR
for boards without plug\-and\-play cabability\&.
.PP
\fB\-i, \-\-irq \fR\fB\fInumber\fR\fR
.PP
Specify irq line
\fInumber\fR
for boards without plug\-and\-play cabability\&.
.PP
\fB\-m, \-\-minor \fR\fB\fInumber\fR\fR
.PP
Configure gpib device file with minor number
\fInumber\fR
(default is 0)\&.
.PP
\fB\-o, \-\-offline\fR
.PP
Unconfigure an already configured board, don\*(Aqt bring board online\&.
.PP
\fB\-p, \-\-pad \fR\fB\fInumber\fR\fR
.PP
Specify primary gpib address\&.
\fInumber\fR
should be in the range 0 through 30\&.
.PP
\fB\-u, \-\-pci\-bus \fR\fB\fInumber\fR\fR
.PP
Specify pci bus
\fInumber\fR
to select a specific pci board\&. If used, you must also specify the pci slot with
\fB\-\-pci\-slot\fR\&.
.PP
\fB\-l, \-\-pci\-slot \fR\fB\fInumber\fR\fR
.PP
Specify pci slot
\fInumber\fR
to select a specific pci board\&. If used, you must also specify the pci bus with
\fB\-\-pci\-bus\fR\&.
.PP
\fB\-s, \-\-sad \fR\fB\fInumber\fR\fR
.PP
Specify secondary gpib address\&.
\fInumber\fR
should be 0 (disabled) or in the range 96 through 127 (0x60 through 0x7f hexadecimal)\&.
.PP
\fB\-\-[no\-]sre\fR
.PP
Assert (or not) remote enable line after bringing board online\&. Default is
\fB\-\-sre\fR\&.
.PP
\fB\-a, \-\-sysfs\-device\-path \fR\fB\fIdev_path\fR\fR
.PP
Select a specific board to attach by its sysfs device path\&. The sysfs device path is the absolute path to the device\*(Aqs directory under /sys/devices, with the leading "/sys" stripped off\&. The device path is available in udev scripts as the DEVPATH variable\&.
.PP
\fB\-\-[no\-]system\-controller\fR
.PP
Configure board as system controller (or not)\&.
.PP
\fB\-v, \-\-version \fR
.PP
Prints the current linux\-gpib version and exits\&.
.SH "COPYRIGHT"
.br
Copyright \(co 2003-2006, 2008 Frank Mori Hess
.br


ChangeLog
2018-12-25 09:50  dpenkler

	* [r1806] Fix board type

2018-12-25 09:46  dpenkler

	* [r1805] Fix ni_usb_gpib cut & paste error

2018-12-25 09:42  dpenkler

	* [r1804] Simplify usb udev rules as per Frank's suggestion

2018-12-21 10:44  dpenkler

	* [r1803] Changes for multiple board configuration

	  Update INSTALL
	  Add serial to device environment
	  Print serial in gpib_udev_config
	  Add example rules for agilent

2018-12-15 08:55  fmhess

	* [r1802] Add some examples for configuring multiple boards with
	  different minor numbers.

2018-12-15 08:29  fmhess

	* [r1801] Automatically gibe the correct board type to gpib_config
	  in udev rules

2018-12-15 08:01  fmhess

	* [r1800] rename gpib generic rules so it runs at 98
	  (before the driver-specific fules which run at 99)

2018-12-15 07:52  fmhess

	* [r1799] Add fix for mylog from gpib_udev_config to
	  gpib_udev_fxloader.in. Also add some quoting to
	  ensure whitespace is preserved in log messages.

2018-12-14 16:23  dpenkler

	* [r1798] Fix logging in gpib_udev_config for gpib_config options
	  --

2018-12-14 08:30  fmhess

	* [r1797] Removed some bash-isms so scripts work with posix
	  /bin/sh.

2018-12-14 08:11  fmhess

	* [r1796] Revert commit 1793

2018-12-14 08:07  fmhess

	* [r1795] Add --with-udev-libdir option to configure, that I failed
	  to include in earlier commit.

2018-12-13 17:05  dpenkler

	* [r1794] Really remove unused files

2018-12-13 17:02  dpenkler

	* [r1793] Add UDEV_LIBDIR to Makefile.ams\nRemove unused files

2018-12-13 13:44  fmhess

	* [r1792] Added --with-udev-libdir configure option, defaulted
	  to PREFIX/lib/udev.

2018-12-13 12:24  dpenkler

	* [r1791] Rename and move udevadm wrapper to libdir

2018-12-13 12:17  fmhess

	* [r1790] Add missing define of UDEV_LIB_DIR to prevent files from
	  getting installed into root directory.

2018-12-13 12:13  fmhess

	* [r1789] Replaced udev scripts ni_usb_gpib and agilent_82357a with
	  a single generic script.

2018-12-13 11:53  fmhess

	* [r1788] Really delete usermap files this time.

2018-12-13 11:47  fmhess

	* [r1787] Pull out firmware loading into a single udev script
	  gpib_udev_fxloader.

2018-12-13 11:44  dpenkler

	* [r1786] Use #!/bin/sh instead of #!/bin/bash

2018-12-13 11:11  dpenkler

	* [r1785] Add wrapper script for udevadm

2018-12-12 18:04  dpenkler

	* [r1784] Fix udev rules scripts for agilent usb

2018-12-12 15:43  fmhess

	* [r1783] Generate usb script/rules during "make all" rather
	  than possibly waiting until "make install", where the
	  user may be running as root.

2018-12-12 15:27  fmhess

	* [r1782] Fixed assignment that was supposed to be a comparison

2018-12-12 14:03  fmhess

	* [r1781] Removed obsolete hotplug usermaps.

	  Fixed problems with usb udev rules not working if the driver
	  module was not already loaded when the hardware was plugged
	  in.

	  Made usb drivers respond to the sysfs device path of their
	  usb interface rather than the parent usb device. This is helpful
	  since the driver binds to the interfaces rather than the device.

2018-12-12 06:25  fmhess

	* [r1780] Update comments in agilent_82357a udev script to use
	  new --serial and --sysfs-device-path gpib_config options.

2018-12-10 10:49  fmhess

	* [r1779] Added linux/atomic.h compatibility header.

2018-12-10 10:36  fmhess

	* [r1778] Added a couple more devices to ni_usb_gpib.usermap.

2018-12-10 10:28  fmhess

	* [r1777] Add kernel driver support for selecting hardware by sysfs
	  device path
	  or serial number.

2018-12-10 10:27  fmhess

	* [r1776] Add support for selecting hardware by sysfs device path
	  or serial number. Updated gpib_config docs with
	  missing info on some options. Noted in docs that
	  gpib.conf will be in the sysconfdir, which isn't
	  always /etc/ depending on configuration.

2018-12-10 10:11  fmhess

	* [r1775] Add device ids for various ni adapters to
	  ni_usb_gpib.usermap

2018-12-07 09:12  dpenkler

	* [r1774] Remove dependency on autoconf tools
	  Modify INSTALL instructions
	  Update drivers/gpib/Makefile
	  Add Makefile to invoke kernel build scripts
	  Add version.h to drivers/gpib/include
	  Make definitions of smp_mb__xxx_atomic independent of kernel
	  version
	  Add VERBOSE flag for make, default is VERBOSE=0
	  so that build log is more overseeable.
	  Add HAVE_DEV_OF_NODE workaround for CentOS/RHEL7
	  Fix compile warnings in lpvo

2018-12-06 06:10  fmhess

	* [r1773] Reverted r1766 since it was actually comparing the serial
	  number to the device tree path. Linux doesn't seem to
	  have a universal (for all bus types) way of specifying
	  a piece of hardware by path, so we can't use the
	  device tree path to specify a usb device anyways.

2018-12-05 11:19  dpenkler

	* [r1772] Apply gpib_lines patch from Tomislav Ivek

2018-11-29 08:29  dpenkler

	* [r1771] Change barrier() to mb() to be generic

2018-11-28 19:02  dpenkler

	* [r1770] Add kernel version ifdef for smp atomics

2018-11-28 18:02  dpenkler

	* [r1769] Add compat fixes for kernels < 3.15 and 3.16

2018-11-28 14:36  dpenkler

	* [r1768] Add patch from Tomislav Ivek

	  The current linux-gpib Python class can run into a double-close
	  situation. This bug is related to a design point in Gpib.Gpib
	  which
	  cleans up its GPIB handle only on Python's GC cycle. This is not
	  guaranteed to run when the programmer expects it to and in
	  extreme
	  cases might even leak resources. This is resolved by adding a
	  Gpib.close() method for user-controlled cleanup.

2018-11-28 10:26  fmhess

	* [r1767] Applied patch from Derek Kozel <derek.kozel@gmail.com>:

	  During the linux-gpib-kernel make install process the kernel
	  modules_install command is run bit the DESTDIR argument is not
	  passed
	  in meaning that it will always try to install to the system
	  directory
	  rather than a custom directory. Adding
	  INSTALL_MOD_PATH="$(DESTDIR)" to
	  the command correctly passes the information.

2018-11-16 20:30  fmhess


linux-gpib-user/lib/gpib_config/gpib_config.c
/***************************************************************************
                              gpib_config.c
                             -------------------

    copyright            : (C) 2001,2002,2003 by Frank Mori Hess
    email                : fmhess@users.sourceforge.net
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/
#define _GNU_SOURCE

#include <assert.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/sysmacros.h>
#include <sys/stat.h>
#include <unistd.h>
#include <getopt.h>
#include "ib_internal.h"

typedef struct
{
	char *config_file;
	char *device_file;
	unsigned int minor;
	char *board_type;
	int irq;
	unsigned long iobase;
	int dma;
	int pci_bus;
	int pci_slot;
	int pad;
	int sad;
	int assert_ifc;
	int assert_remote_enable;
	int offline;
	int is_system_controller;
	void *init_data;
	int init_data_length;
	char *sysfs_device_path;
} parsed_options_t;

static void help(void)
{
	printf("gpib_config [options] - configures a GPIB interface board\n");
	printf("\t-t, --board-type BOARD_TYPE\n"
		"\t\tSet board type to BOARD_TYPE.\n");
	printf("\t-c, --device-file FILE_PATH\n"
		"\t\tSpecify character device file path for the board.\n"
		"\t\tThis can be used as an alternative to the --minor option.\n");
	printf("\t-d, --dma NUM\n"
		"\t\tSpecify isa dma channel NUM for boards without plug-and-play cabability.\n");
	printf("\t-I, --init-data FILE_PATH\n"
		"\t\tSpecify file containing binary initialization data (firmware) for board.\n");
	printf("\t--[no-]ifc\n"
		"\t\tPerform (or not) interface clear after bringing board online.  Default is --ifc.\n");
	printf("\t-b, --iobase NUM\n"
		"\t\tSet io base address to NUM for boards without plug-and-play cabability.\n");
	printf("\t-i, --irq NUM\n"
		"\t\tSpecify irq line NUM for boards without plug-and-play cabability.\n");
	printf("\t-f, --file FILE_PATH\n"
		"\t\tSpecify file path for configuration file.  The values in the configuration\n"
		"\t\tfile will be used as defaults for unspecified options.  The default configuration\n"
		"\t\tfile is %s\n", DEFAULT_CONFIG_FILE);
	printf("\t-h, --help\n"
		"\t\tPrint this help and exit.\n");
	printf("\t-m, --minor NUM\n"
		"\t\tConfigure gpib device file with minor number NUM (default 0).\n"
		"\t\tAlternatively, the device file may be specified with the --device-file option.\n");
	printf("\t-o, --offline\n"
		"\t\tUnconfigure an already configured board, don't bring board online.\n");
	printf("\t-p, --pad NUM\n"
		"\t\tSpecify primary gpib address.  NUM should be in the range 0 through 30.\n");
	printf("\t-u, --pci-bus NUM\n"
		"\t\tSpecify pci bus NUM to select a specific pci board.\n"
		"\t\tIf used, you must also specify the pci slot with --pci-slot.\n");
	printf("\t-l, --pci-slot NUM\n"
		"\t\tSpecify pci slot NUM to select a specific pci board.\n"
		"\t\tIf used, you must also specify the pci bus with --pci-bus.\n");
	printf("\t-s, --sad NUM\n"
		"\t\tSpecify secondary gpib address.  NUM should be 0 (disabled) or in the range\n"
		"\t\t96 through 126 (0x60 through 0x7e hexadecimal).\n");
	printf("\t--[no-]sre\n"
		"\t\tAssert (or not) remote enable line after bringing board online.  Default is --sre.\n");
	printf("\t-a, --sysfs-device-path DEVPATH\n"
		"\t\tSelect a specific board to attach by its sysfs device path.\n");
	printf("\t--[no-]system-controller\n"
		"\t\tConfigure board as system controller (or not).\n");
	printf("\t-v, --version\n"
		"\t\tPrint version of linux-gpib library and exit.\n");
}

static int load_init_data(parsed_options_t *settings, const char *file_path)
{
	int retval;
	FILE *init_file = fopen(file_path, "r");
	if (init_file)	{
		struct stat file_status;
		if (fstat(fileno(init_file), &file_status) == 0) {
			settings->init_data = malloc(file_status.st_size);
			if (settings->init_data) {
				settings->init_data_length = fread(settings->init_data, 1, file_status.st_size, init_file);
				if (settings->init_data_length == file_status.st_size)
					retval = 0;
				else {
					settings->init_data_length = 0;
					free(settings->init_data);
					settings->init_data = NULL;
					fprintf(stderr, "fread() returned short read\n");
					retval = -EIO;
				}
			} else {
				fprintf(stderr, "malloc() failed.\n");
				perror(__FUNCTION__);
				retval = -errno;
			}
		} else {
			fprintf(stderr, "fstat() failed on file \'%s\'.\n", file_path);
			perror(__FUNCTION__);
			retval = -errno;
		}
	} else {
		fprintf(stderr, "Failed to open file \'%s\' for reading.\n", file_path);
		perror(__FUNCTION__);
		retval = -errno;
	}
	fclose(init_file);
	return retval;
}

static int parse_options(int argc, char *argv[], parsed_options_t *settings)
{
	int c, index;
	int retval;
	char *version;
	struct option options[] =
	{
		{ "iobase", required_argument, NULL, 'b' },
		{ "device-file", required_argument, NULL, 'c' },
		{ "sysfs-device-path", required_argument, NULL, 'a' },
		{ "dma", required_argument, NULL, 'd' },
		{ "file", required_argument, NULL, 'f' },
		{ "help", no_argument, NULL, 'h' },
		{ "init-data", required_argument, NULL, 'I' },
		{ "irq", required_argument, NULL, 'i' },
		{ "pci-slot", required_argument, NULL, 'l' },
		{ "minor", required_argument, NULL, 'm' },
		{ "offline", no_argument, NULL, 'o' },
		{ "pad", required_argument, NULL, 'p' },
		{ "sad", required_argument, NULL, 's' },
		{ "board-type", required_argument, NULL, 't' },
		{ "pci-bus", required_argument, NULL, 'u' },
		{ "version", no_argument, NULL, 'v' },
		{ "no-ifc", no_argument, &settings->assert_ifc, 0 },
		{ "ifc", no_argument, &settings->assert_ifc, 1 },
		{ "no-sre", no_argument, &settings->assert_remote_enable, 0 },
		{ "sre", no_argument, &settings->assert_remote_enable, 1 },
		{ "no-system-controller", no_argument, &settings->is_system_controller, 0 },
		{ "system-controller", no_argument, &settings->is_system_controller, 1 },
		{ 0 },
	};

	memset(settings, 0, sizeof(parsed_options_t));
	settings->irq = -1;
	settings->iobase = (unsigned long)-1;
	settings->dma = -1;
	settings->pci_bus = -1;
	settings->pci_slot = -1;
	settings->pad = -1;
	settings->sad = -1;
	settings->assert_ifc = 1;
	settings->assert_remote_enable = 1;
	settings->is_system_controller = -1;

	while (1) {
		c = getopt_long(argc, argv, "a:b:c:d:e:f:hi:I:l:m:op:s:t:u:v", options, &index);
		if (c == -1)
			break;
		switch(c) {
		case 0:
			break;
		case 'a':
			free(settings->sysfs_device_path);
			settings->sysfs_device_path = strdup(optarg);
			break;
		case 'b':
			settings->iobase = strtol(optarg, NULL, 0);
			break;
		case 'c' :
			free(settings->config_file);
			settings->device_file = strdup(optarg);
			break;
		case 'd':
			settings->dma = strtol(optarg, NULL, 0);
			break;
		case 'f':
			free(settings->config_file);
			settings->config_file = strdup(optarg);
			break;
		case 'h':
			help();
			exit(0);
			break;
		case 'I':
			retval = load_init_data(settings, optarg);
			if (retval < 0)
				return retval;
			break;
		case 'i':
			settings->irq = strtol(optarg, NULL, 0);
			break;
		case 'l':
			settings->pci_slot = strtol(optarg, NULL, 0);
			break;
		case 'm':
			settings->minor = strtol(optarg, NULL, 0);
			break;
		case 'o':
			settings->offline = 1;
			break;
		case 'p':
			settings->pad = strtol(optarg, NULL, 0);
			break;
		case 's':
			settings->sad = strtol(optarg, NULL, 0);
			break;
		case 't':
			free(settings->board_type);
			settings->board_type = strdup(optarg);
			break;
		case 'u':
			settings->pci_bus = strtol(optarg, NULL, 0);
			break;
		case 'v':
		        ibvers(&version);
			printf("linux-gpib version = %s\n",version);
			exit(0);
			break;
		default:
			help();
			exit(1);
		}
	}
	if (settings->device_file) {
		struct stat file_stats;
		if (stat(settings->device_file, &file_stats) < 0) {
			fprintf(stderr, "Failed to get file information on file \"%s\".\n", settings->device_file);
			perror(__FUNCTION__);
			return -errno;
		}
		if (S_ISCHR(file_stats.st_mode) == 0) {
			fprintf(stderr, "The device file \"%s\" is not a character device.\n", settings->device_file);
			return -EINVAL;
		}
		settings->minor = minor(file_stats.st_rdev);
	} else {
		if (asprintf(&settings->device_file , "/dev/gpib%i", settings->minor) < 0)
			return -ENOMEM;
	}
	return 0;
}

static int configure_sysfs_device_path(int fileno, const char *sysfs_device_path)
{
	struct gpib_select_device_path_ioctl devpath_selection;
	int retval;

	if (sysfs_device_path != NULL) {
		if (strlen(sysfs_device_path) >= sizeof(devpath_selection.device_path))	{
			fprintf(stderr, "device path too long.\n");
			return -EINVAL;
		}
		strncpy(devpath_selection.device_path, sysfs_device_path,
			sizeof(devpath_selection.device_path));
	} else {
		memset(devpath_selection.device_path, 0, sizeof(devpath_selection.device_path));
	}
	retval = ioctl(fileno, IBSELECT_DEVICE_PATH, &devpath_selection);
	if (retval < 0)	{
		/* If the user didn't request any device path, EINVAL error is probably just
		 * due to using an older kernel module that doesn't support this ioctl.  So
		 * only error out if a path was specified. */
		if (errno != EINVAL || strlen(devpath_selection.device_path) > 0) {
			fprintf(stderr, "failed to configure device path \"%s\"\n", devpath_selection.device_path);
			return retval;
		}
	}
	return 0;
}

static int configure_board(int fileno, const parsed_options_t *options)
{
	struct gpib_board_type_ioctl boardtype;
	struct gpib_select_pci_ioctl pci_selection;
	struct gpib_pad_ioctl pad_cmd;
	struct gpib_sad_ioctl sad_cmd;
	struct gpib_online_ioctl online_cmd;
	int retval;

	online_cmd.online = 0;
	online_cmd.init_data_ptr = 0;
	online_cmd.init_data_length = 0;
	retval = ioctl(fileno, IBONL, &online_cmd);

	if (retval < 0)	{
		fprintf(stderr, "failed to bring board offline\n");
		return retval;
	}
	if (options->offline != 0)
		return 0;
	strncpy(boardtype.name, options->board_type, sizeof(boardtype.name) - 1);
	retval = ioctl(fileno, CFCBOARDTYPE, &boardtype);
	if (retval < 0)	{
		fprintf(stderr, "failed to configure boardtype: %s\n", boardtype.name);
		return retval;
	}
	retval = ioctl(fileno, CFCBASE, &options->iobase);
	if (retval < 0)	{
		fprintf(stderr, "failed to configure base address\n");
		return retval;
	}
	retval = ioctl(fileno, CFCIRQ, &options->irq);
	if (retval < 0)	{
		fprintf(stderr, "failed to configure irq\n");
		return retval;
	}
	retval = ioctl(fileno, CFCDMA, &options->dma);
	if (retval < 0)	{
		fprintf(stderr, "failed to configure dma channel\n");
		return retval;
	}
	pad_cmd.handle = 0;
	pad_cmd.pad = options->pad;
	retval = ioctl(fileno, IBPAD, &pad_cmd);
	if (retval < 0)	{
		fprintf(stderr, "failed to configure pad\n");
		return retval;
	}
	sad_cmd.handle = 0;
	sad_cmd.sad = options->sad;
	retval = ioctl(fileno, IBSAD, &sad_cmd);
	if (retval < 0)	{
		fprintf(stderr, "failed to configure sad\n");
		return retval;
	}
	pci_selection.pci_bus = options->pci_bus;
	pci_selection.pci_slot = options->pci_slot;
	retval = ioctl(fileno, IBSELECT_PCI, &pci_selection);
	if (retval < 0)	{
		fprintf(stderr, "failed to configure pci bus\n");
		return retval;
	}

	configure_sysfs_device_path(fileno, options->sysfs_device_path);

	online_cmd.online = 1;
	assert(sizeof(options->init_data) <= sizeof(online_cmd.init_data_ptr));
	online_cmd.init_data_ptr = (uintptr_t)options->init_data;
	online_cmd.init_data_length = options->init_data_length;
	retval = ioctl(fileno, IBONL, &online_cmd);
	if (retval < 0)	{
		fprintf(stderr, "failed to bring board online\n");
		return retval;
	}

	retval = ibconfig(options->minor, IbcSRE, options->assert_remote_enable);
	if (retval < 0)	{
		fprintf(stderr, "ibconfig IbcSRE %d failed\n", options->assert_remote_enable);
	}

	retval = ibrsc(options->minor, options->is_system_controller);
	if (retval & ERR) {
		fprintf(stderr, "failed to request/release system control\n");
		return -1;
	}
	if (options->is_system_controller) {
		if (options->assert_ifc) {
			retval = ibsic(options->minor);
			if (retval & ERR) {
				fprintf(stderr, "failed to assert interface clear\n");
				return -1;
			}
		}
	}

	return 0;
}

int main(int argc, char *argv[])
{
	ibConf_t configs[FIND_CONFIGS_LENGTH];
	ibBoard_t boards[GPIB_MAX_NUM_BOARDS];
	char *filename, *envptr;
	int retval;
	parsed_options_t options;
	ibBoard_t *board;
	ibConf_t *conf = NULL;
	int i;

	retval = parse_options(argc, argv, &options);
	if (retval < 0)	{
		fprintf(stderr, "failed to parse command line options.");
		return retval;
	};

	envptr = getenv("IB_CONFIG");
	if (options.config_file) filename = options.config_file;
	else if (envptr) filename = envptr;
	else filename = DEFAULT_CONFIG_FILE;

	if (options.minor >= GPIB_MAX_NUM_BOARDS) {
		fprintf(stderr, "minor number %i out of range\n", options.minor);
		return -1;
	}

	retval = parse_gpib_conf(filename, configs, FIND_CONFIGS_LENGTH,
				 boards, GPIB_MAX_NUM_BOARDS, options.minor);
	if (retval < 0)	{
//              Message printed in parser
//		fprintf(stderr, "failed to parse config file %s\n", filename);
		return retval;
	}


	for (i = 0;i < FIND_CONFIGS_LENGTH;i++) {
		if (configs[i].is_interface == 0)
			continue;
		if (configs[i].settings.board != options.minor)
			continue;
		conf = &configs[i];
		break;
	}

	if (i == FIND_CONFIGS_LENGTH) {
		fprintf(stderr, "Could not find config for minor %d\n", options.minor);
		return -1;
	}

	board = &boards[options.minor];

	if (options.board_type == NULL) {
		options.board_type = strdup(board->board_type);
		if (options.board_type == NULL)
			return -ENOMEM;
	}

	if ((strlen(options.board_type) == 0) && !options.offline) {
		fprintf(stderr, "No board type configured for minor %d\n", options.minor);
		return -1;
	}

	if (options.irq < 0)
		options.irq = board->irq;
	if (options.iobase == (unsigned long)-1)
		options.iobase = board->base;
	if (options.dma < 0)
		options.dma = board->dma;
	if (options.pci_bus < 0)
		options.pci_bus = board->pci_bus;
	if (options.pci_slot < 0)
		options.pci_slot = board->pci_slot;

	if (options.pad < 0)
		options.pad = conf->settings.pad;
	if (!options.offline && (options.pad < 0 || options.pad > 30)) {
		fprintf(stderr, "Invalid or no pad %d configured for minor %d\n", options.pad, options.minor);
		return -1;
	}

	if (options.sad == -1) { // not specified
		options.sad = conf->settings.sad;
	} else {
		if (options.sad  != 0 && options.sad < 96 || options.sad > 127) {
			fprintf(stderr, "Invalid sad %d configured for minor %d\n", options.sad, options.minor);
			return -1;
		}
		if (options.sad == 0)
			options.sad = -1;
		else
			options.sad -= sad_offset;
	}

	if (options.is_system_controller < 0)
		options.is_system_controller = board->is_system_controller;

	board->fileno = open(options.device_file, O_RDWR);
	if (board->fileno < 0) {
		fprintf(stderr, "failed to open device file '%s'\n", options.device_file);
		perror(__FUNCTION__);
		return board->fileno;
	}
	if (options.sysfs_device_path == NULL) {
		if (strlen(board->sysfs_device_path) > 0) {
			options.sysfs_device_path = strdup(board->sysfs_device_path);
			if (options.sysfs_device_path == NULL)
				return -ENOMEM;
		}
	}
	retval = configure_board(board->fileno, &options);
	if (retval < 0) {
		fprintf(stderr, "failed to configure board\n");
		perror(__FUNCTION__);
		return retval;
	}
	close(board->fileno);
	free(options.init_data);
	free(options.device_file);
	free(options.config_file);
	free(options.board_type);

	return 0;
}
